mautic/mautic
# Device granularity

Please answer the following questions. 

| Q  | A
| --- | ---
| Bug fix? | y
| New feature? | y
| Related user documentation PR URL | (https://github.com/mautic/documentation/pull/70)
| Related developer documentation PR URL | n
| Issues addressed (#s or URLs) | #2045
| BC breaks? | n
| Deprecations? | n

**Note that all new features should have a related user and/or developer documentation PR in their respective repositories.** 

### Required
#### Description:
This PR adds new device granularity to page hits. A new device granularity on lifecycle dashboard and a new devices widget for page hits.

Email open events will also track devices used to open emails, a contact's devices report will be added to the reports section

### To Test
- Apply PR
- Run composer update
- Migrate db
- visit a page and open an email using various devices
- you should see the contact's timeline show what devices have been used, and also the charts on email and widgets.
-report for page hits can also add a column for hit devices
